### ISSUE TITLE
more component discussion

### ISSUE THREAD

**Issue by jdegoes** [slamdata/purescript-halogen#84]
I spent part of the weekend convincing myself the types for component are wrong. Components are a black hole, it's not possible to interrogate them (for example, asking an Ace component what the current contents of the buffer are). While components emit declarative information in response to every input, that's not useful for communication; they cannot yield event information except in response to user input.

I think the signature I'm really looking for is something like this:

```purescript
newtype Component p m i req res = Component (SF1 (Either i req) (Tuple (HTML p (m i)) (m res)))

type ComponentW p m req res = ComponentW (exists i. Component p m i req res)
```

This is possibly not exactly right, one can imagine wanting the `Tuple` to be an `Either`, but that's not precise enough either and allows a component to refuse to emit `HTML`. OTOH encoding more strictly reduces the power (e.g. prevents emitting `res` from user input or prevents updating HTML in response to requests).

**Comment by puffnfresh** [comment]
For people still new, like me, the current Component looks like:

```purescript
newtype ComponentF p m req res i = ComponentF (SF1 (Either i req) (HTML p (m (Either i res))))

newtype Component p m req res = Component (Exists (ComponentF p m req res))
```

**Comment by paf31** [comment]
> Components are a black hole, it's not possible to interrogate them

I thought that was by design - you don't get to interrogate, only observe what messages are exposed.

If you use `Tuple` as in your first example, you won't be able to send non-external messages on user input, and if you use `Either`, you don't necessarily ever get a document to render.

What's the problem we're trying to solve?

**Comment by paf31** [comment]
I do think there is a problem with `Component` though, which is that they aren't as composable as raw signal functions. I think the existential gets in the way. It makes it hard to have things like lists of components, because you can't construct the appropriate existentially hidden type in any sensible way to combine them.

I think a good step would be to make the internal message type into a type variable, instead of quantifying it. Then we could write combinators like

```purescript
list :: forall p m i res req. List (Component p m i res req) -> Component p m (Tuple Int i) (Tuple Int res) (Tuple Int req)
``` 

Either that, or just get rid of the `Component` abstraction entirely, since at that point we're basically back to a newtype on top of raw signal functions anyway.

**Comment by jdegoes** [comment]
> What's the problem we're trying to solve?

It's quite simple: take the hypothetical Ace component as an example. Right now, you can build a toolbar which sends commands to the Ace component, as we have done in examples. But if you want to, for example, retrieve the content of the editor, or do any one of numerous other things, you have to go outside the component interface, or send something like `AVar` to a component, which will lead to very imperative code.

Components need a way not just to receive information, but to send it, more than that, to reply to requests for information. That's the only way a component can be completely encapsulated.

On the whole existential thing, I think the component's wrong type signature is really what's screwing it up here. Ideally, there would be two forms, a raw component:

```purescript
newtype Component p m i req res = Component (SF1 (Either i req) (Tuple (HTML p (m i)) (m res)))
```

and one with existential hiding which is much easier to compose:

```purescript
type ComponentW p m req res = ComponentW (exists i. Component p m i req res)
```

because you don't have to worry about `i`, useful when you want to crush the entropy that otherwise accumulates and seal off the possibility of driving a component with fake inputs.

> If you use Tuple as in your first example, you won't be able to send non-external messages on user input

Well, user input can affect the state of the component, which can in turn yield `res` values. But yeah, I agree the type signature is not exactly right, at best it's an approximation.

Part of the problem is that ideally the component can respond to input or respond to requests, but is constrained by the fact that it must continuously produce a model of the HTML (indeed, this is necessary since one can imagine many requests updating the HTML).

The point is that this signature is wrong:

```purescript
newtype Component p m req res = Component (SF1 req (Tuple (HTML p (m res))))
```

...precisely because it creates a black hole. One cannot obtain any data about the state of the component, which prevents creation of a useful Ace component, and also prevents our use case for Cells being Components which communicate to the container via a well-defined ADT protocol.

**Comment by paf31** [comment]
As I say, I agree that something isn't quite right, but this would be a massive departure from what we've already done. At the start of the project, I suggested using explicit reactive variables, and passing them around, allowing us to potentially create data dependency graphs with cycles, but we decided the approach was too imperative. That's the reason I liked the signal function abstraction - there's a really clear, pure interpretation, and you can still represent cycles by tying the knot at the top level. 

In fact, I'm not convinced we _can't_ do what you want to do with the cell component - why do you think it's impossible? Even with the Ace example, you can keep the current content as state if you like, just by emitting events when the content changes. I guess that without more details about why you think things don't work, it'll be hard to make progress on this.

I'm not sure if adding another type of `Component` is the best option. We already have `SF` and `Component`, and `ComponentW` can be easily implemented in terms of `Component` above, by setting `i` to be `Void`.

Is the problem that you want the signal functions to be more like bidirectional channels? If so, I'd like to try to convince you that we don't need them, since the tying-the-knot trick essentially gives us bidirectionality.

**Comment by paf31** [comment]
> The point is that this signature is wrong precisely because it creates a black hole

That is by design though, right? We didn't want arbitrary access to state. `req` was there to control what writes you were allowed to request, and `res` was there to control what information about the state you wanted to leak to the outside.

One approach might be to unroll the nu-type in `SF1`:

```purescript
newtype SF1 s i o = SF1 { state :: s, step :: s -> i -> Tuple o s }
```

but I'm not sure that buys us much, or that it's a good thing to make the state visible.

**Comment by jdegoes** [comment]
>  It makes it hard to have things like lists of components, because you can't construct the appropriate existentially hidden type in any sensible way to combine them.

What if components had some notion of identity, such that you could do routing like that? It'd have to be done in FFI or (carefully inspected and tested) unsafe PS, but might be possible.

> req was there to control what writes you were allowed to request, and res was there to control what information about the state you wanted to leak to the outside.

The problem is you cannot emit `res` whenever you want. Rather, `req` can affect state, which can affect view, and can therefore affect `res`, but `req` cannot be responded to with `res`.

Let's take the Ace component example. I want a toolbar that takes the current selection and makes it bold. The Ace component doesn't know anything about that (it would be conflation of concerns). Currently, I can send some commands to Ace to insert text at the current position -- but what I really what is to get the current selection, so I can apply my bold formatting to it, and then overwrite the current selection with the text.

The protocol supported by this component might look something like this:

```purescript
data AceReq = InsertText Int String | GetSelection | ...

data AceRes = Selection Int Int | ...
```

This provides a way for me to get information out of the component *on-demand*. The alternative is to provide *all possible information* as events, so that parents can *reconstruct the state* from a monoid on the events. That requires component-users to know enough about the way a component works to be able to recreate its internal state tracking behavior by monoidally appending its events. This does not seem like it's an example of good design.

Also widgets cannot necessarily be made to produce all the events which can reconstruct the internal state of the component, making the strategy only usable for pure Halogen components.

In SlamData we want to use components for cells, which presents many of the same challenges.

Unrolling the nu-type is in some sense equivalent to allowing requests to return responses, but only for a pure Halogen component; for a widget the state may be stored in the DOM.

>  That's the reason I liked the signal function abstraction

I'm not suggesting an impure abstraction, only a pure one that allows components to selectively emit information. Right now `req` is always equal to `res` because that's the only thing that makes sense, so you have this invariant functor in which both requests and responses must share the same ADT (which is a bit like taking a function `a -> b` and turning it into `Either a b -> Either a b` just so you can make it `i -> i` for some `i`). While that makes the functor invariant and limits composition, the bigger problem is that the component may emit no information except on user input events (or initializers / finalizers), which does not permit "state interrogation" and creates barriers for encapsulation.

> I think a good step would be to make the internal message type into a type variable, instead of quantifying it. Then we could write combinators like

These sorts of solutions work good for small examples, but remember that you cannot ever throw away any information, so the tiniest little internal details of some component deeply embedded into some button of a widget will propagate by sheer brute force to the top of the application. I think that's some of the complexity that @garyb is struggling with right now.

**Comment by paf31** [comment]
> Let's take the Ace component example. I want a toolbar that takes the current selection and makes it bold. The Ace component doesn't know anything about that (it would be conflation of concerns). Currently, I can send some commands to Ace to insert text at the current position -- but what I really what is to get the current selection, so I can apply my bold formatting to it, and then overwrite the current selection with the text.

The way I figured this would work would be to have a single input to the Ace component like

```purescript
data Input = MakeBold (Maybe Selection) | ...
```

Really, it sounds like you want your inputs to be commands with return types. Effectively they all have return type `Unit` right now.

I was thinking about this a bit this morning, and about @garyb's `Inject` idea too, and I wonder if something like this would work:

```purescript
newtype SF f i o  = SF (i -> f (SF1 f i o))

newtype SF1 f i o = SF1 o (SF f i o)

newtype Component f p m req res = Component (exists i. SF1 (FreeC f) (Either i req) (HTML p (m (Either i res))))
```

Quite a mouthful, but maybe we can break it up.

The idea is that if you want the internal structure of your component to support commands like `getSelection` which returns a `Selection`, you build some free monad which describes those commands, and then bake that into your signal function. So this is not to be thought of as _effectful signal functions_, but signal functions in which we support some internal commands. For the vast majority of cases, I imagine `f` will be trivial.

To interpret this, you need a natural transformation taking `FreeC f` to something like `Eff`.

To combine components, you take the free monad coproduct, and get to use `Inject` to compose things.

I think this might be overkill, personally, but it might work, and you still keep the state internal to the component - you just get a better language with which you can compose commands - which means your commands can be finer grained and actually isolate things better.

**Comment by jdegoes** [comment]
> Really, it sounds like you want your inputs to be commands with return types.

Yes, session typing is really what I want but that's impossible in PureScript (at least in any usable fashion, AFAIK).

> you build some free monad which describes those commands, and then bake that into your signal function

I think this approach has a lot promise and we should think about it more seriously. I need to inspect your proposed types more to understand the power and limitations of the approach.

In retrospect my example of a "function on state" is not exactly the best example because of this encoding:

```purescript
data Selection = Selection Int Int String
data AceReq = GetSelection (Selection -> AceReq) | SetSelection Selection | ...
```

But as a protocol it still conflates request type and response types, leads to the ever-present invariant functor, etc. Also doesn't really avoid the problem: how can I tell the component to give me its JSON serialization so I can embed it in more JSON? I can't, my "function on state" can only issue commands the component already understands. It's broken.

We do need to hammer this out and relatively soon. I'll spend more time on the free approach.

**Comment by paf31** [comment]
I think I'm overthinking the `Free` approach. If we want the requests to be able to support a rich language of terms with things like `getSelection`, maybe we can just use an input type of `Free f Unit` for some `f`. I think that's sort of what you're saying above, right? We keep the "black box" property of the components, but get a nice language with which to make requests.

On the one hand, you have a simple ADT of inputs - not very expressive, but controlled. On the other hand, you have something like `s -> Node -> Eff eff Unit` - very expressive, but completely uncontrolled. I think a free monad could be a nice compromise.

**Comment by garyb** [comment]
I quite like the sound of using `Free` for this too. I've not followed the fine details of the discussion enough to have much of an idea on whether it fully solves the problem, but in principle it certainly seems promising.

**Comment by paf31** [comment]
Another thing I've been thinking about is how to share between components. Right now, there isn't really a good option, you have to just build one big component. One way might be to expose the state type in the signal function:

```purescript
newtype SF1 s i o = SF1 { state :: s, step :: s -> i -> Tuple o s }
```

This would let us write a whole bunch of new combinators (including using lenses for composition), but I think we lose `Apply` (in the current version, we tuple the state types in `<*>`).

One option could be to split the rendering function out:

```purescript
newtype Render s o = Render (s -> o)

newtype SF s i = SF s (s -> i -> s)
```

(this is starting to look quite a bit like `thermite` actually), so that you build your rendering function using things like `Applicative`, and maybe lenses too, and build your signal function separately. This way, different components can share parts of the state, but have their own parts too.

**Comment by jdegoes** [comment]
So here are the following problems with the `Free` approach sketched out above:

1. The monad requires an interpreter, and the `Component` is the only thing that knows how to interpret it. That is, a component is the only thing that knows how to interpret its F-algebra, so it must be the one to do the interpretation, or at least provide a natural transformation to some more generic algebra (or monad).
2. The `Component` interpretation of its F-algebra must be pure or have effects limited to `m`. Actually, the restriction is even more severe, since a `Component` is not allowed to have effects `m`, but it can produce event handlers which produce `m`-effectful events. The distinction is extremely important, because it means that if effects are required to answer a query in `f`, those effects have to be deferred until some arbitrary handler does something interesting in the future. This is nonsense, unfortunately.
3. The `f` input passed to the `Component` cannot represent in any meaningful sense the rest of the program. This means if it's going to have any effect on the world, it will have to do so through effectful, imperative code (it's basically a callback!). So all your input commands to the `Component` must terminate in something like `AVar`'s `putVar`, or `Eff` or `Aff` or something equally effectful.
4. If `Free f Unit` is the input, then `Free f Unit` must be the output, to satisfy the way Halogen wants UIs to work, which not only conflates internal state updating events with external commands, but makes the output useless to anyone else, since in general interpreting the output would require the ability to interpret the F-algebra, something only the `Component` knows how to do.

In summary, the idea of using Free monads to describe DSLs for components sounds attractive, but I have no idea what it would look like in practice. If it can be made to work, I don't think it would be any minor modification of type parameters to SF1. It would be a total rethinking of the approach and I don't know what that would look like.

I took another look at bichannels, and it's not quite right either:

```purescript
type Bichannel a a' b b' f r = Channel (Either a b) (Either a' b') f r
```

One can imagine `a -> a'` as being specialized to `i -> i`, that is, events driven by the user, which makes it invariant with respect to `i`, but which can be existentially hidden (if desired); and `b -> b'` as being `req -> res`. But here channels are allowed to terminate with final result `r`, which is not needed for Halogen (although one can imagine using it for closing dialogs and the like). Also the effect `f` is very powerful, capable of blocking any UI output.

```purescript
newtype SF1 s i o = SF1 { state :: s, step :: s -> i -> Tuple o s }
```

This solves sharing between components, which permits upward propagation of state information (i.e. parent can peek into children's state). It sort of awkwardly requires that children expose answers to all possible queries in part of their state (e.g. the current selection, the line number, the number of lines, the text of the buffer, etc.), which may not yield good performance, especially with 3rd-party widgets -- and assuming its doable at all, which it seems not to be for widgets because that state changes at arbitrary times (from the components point of view).

>  I think we lose Apply (in the current version, we tuple the state types in <*>).

We actually don't need that, we just need `IApply`, i.e. the invariant equivalent, which permits zipping these things together, creating tupled `s`'s that can then be unnested and fed to an iso-function. Need PRs to purescript-invariant. :)

```purescript
newtype Render s o = Render (s -> o)

newtype SF s i = SF s (s -> i -> s)
```

Here are the capabilities I'd like to see:

1. Component can handle input generated from its own event handlers. In theory, this can be existentially hidden, because the values are generated internal to the component and consumed by the component. In practice, sometimes it might be nice to feed a component fake input.
2. Component can respond to requests for information (generally, these are queries about state, but in the case of a widget, whatever state the pure part has is probably incomplete compared to the state that exists outside the pure part).
3. Component can signal parent with new events. I'm thinking, a user clicks on an echart, and the echart notifies the parent component. Something like stackable bichannels here, but it can't be monadic (too powerful).

Exposing the state is orthogonal to (1) and maybe (3), and seems to partially solve (2), at least for pure components (with a big question mark for widgets).

What kind of types let us do that?

1. Component has some initial state `s`.
2. Component has a renderer for that state which produces inputs `i`.
3. Component accepts internal input `i` or external requests `req` and effectfully (?) produces `s -> s` and `res`. 
4. `i` (and `s`) can be existentially hidden and in the case of widgets, there are no `i`s produced or consumed (`Void`), because the state's all hidden; effectively, widgets just accept `req` and produce `res`.

I think approaching it from the point of view of, "What do widgets force us to do?" is helpful. For example, sharing state works for communicating information to parents, but is problematic for widgets. That sort of thinking which uses the common denominator between pure component and widget to extract out a facade that works for both.

One thing separating state seems to allow is for the signal function to not produce the view -- only to produce a state updater (`newtype SF s i = SF i -> (s -> s)`). The reason I like that is because a state updater can be produced inside `Aff` **without blocking the UI**. There are **no race conditions** between state updaters because they don't have any "old" state to capture over.

So one can imagine a component being able to respond to requests and produce state updaters **both inside the same monad**.

```purescript
newtype Component p m req res s i = 
  Component { initial :: s, render :: s -> HTML p i, step :: Either i req -> m (Tuple res (s -> s)) }
```

Although this could probably be cleaned up quite a bit! It's invariant in `i` and `s`, have to think if both those can be hidden & composed with the usual invariant machinery. Here `m` would be `Par eff` and not `Aff eff` because composition should be in parallel and not sequential (also monad may be too powerful).

In this representation, it doesn't matter or what effects transpire to (a) reply to a request, or (b) produce a state updater; the UI is never blocked and there are no state race conditions.

Actually, it's getting late and I've been really sick, so I want to rethink this over the weekend! :smile:

**Comment by paf31** [comment]
> The monad requires an interpreter, and the Component is the only thing that knows how to interpret it

Yes, but I don't see that as a disadvantage. The free monad is _just another data structure_. We already write interpreters for our inputs, using pattern matching.

> If Free f Unit is the input, then Free f Unit must be the output, to satisfy the way Halogen wants UIs to work

Only at the top level. Individual components do not have this restriction, it's just that when you glue your components together, the final component must have `res = req`.

I'm not suggesting the free monad approach as a general framework, just another technique which users can use to structure their code. I wouldn't even include support for it in `halogen`.

I've been thinking about the problem in the following way: before we had `Component`, the library was expressive but things were hard to compose. It is perfectly possible to write a full app using `SF` and `HTML`, composing things by hand. The problem is that it requires a lot of manual plumbing, and the question is, how do we make this less painful? 

`Component` gives one answer, but not a great one in my opinion - it solves one problem, but at a massive price in terms of composability.

I think the better approach is to stick to `SF` and `HTML`, and to provide better tools for composing them.  In most practical code, people don't seem to be using `Component` anyway. 

I think invariance is just something we have to deal with if we want to stick with the signal function/no disappearing information approach. I do think lenses can help, but I don't fully see how yet. Maybe we can restructure things to make lenses more applicable (`thermite` does a decent job of this, I think - see @joneshf 's changes on the todo-mvc example).

---

Another option, which I've thought about, but I don't fully like, is to have a `Component` represent an effectful, self-contained bundle of views and logic, a la React. You would still be able to send messages in, and subscribe to outputs, but diffing would be contained inside the component. This pushes us in the direction of manually-configured graphs of components, which I don't like as much. The global diffing approach is a bit inefficient, perhaps, but allows for some nice abstractions.

**Comment by jdegoes** [comment]
> Yes, but I don't see that as a disadvantage. The free monad is just another data structure. We already write interpreters for our inputs, using pattern matching.

Perhaps I should have made the point more forcefully: the free monad approach cannot work at all in Halogen. The only way to leak information out is through event handlers, which fire only in response to user input. So a free monad can only be used for "write-only" DSLs, which is not necessary since you can just pass (input) parameters to the component.

> Only at the top level. Individual components do not have this restriction, it's just that when you glue your components together, the final component must have res = req.

For all practical intent and purpose, req = res. It's theoretically possible for them to differ but then the SF / component cannot respond to its own input, which is relatively useless. At the very least, res is a superset of req. Also I've never seen real code in which they differ. :smile: 

> I'm not suggesting the free monad approach as a general framework, just another technique which users can use to structure their code. I wouldn't even include support for it in halogen.

Again, the free monad approach is a dead end, there is nothing useful that can be done with them in the current design. One can imagine bundling the interpreter (natural transformation from `f ~> Aff eff` with the component, that would make it useful but extremely imperative. An `m` in a `SF` output can of course be a free monad but cannot be interpreted by the component (so it's really a way of *providing* a DSL to components that they can use, not *exposing* a component-specific DSL to consumers of components).

> The problem is that it requires a lot of manual plumbing, and the question is, how do we make this less painful?

I'm all in favor of this. I think `Inject1` helps and beefing up purescript-invariant would also help quite a lot.

> Another option, which I've thought about, but I don't fully like, is to have a Component represent an effectful, self-contained bundle of views and logic, a la React.

My main drive for this is widgets, i.e. those colossal JS libs that are impossible to replicate in any reasonable amount of time (ace, echarts, etc.). Widgets *by their very nature* bundle MVC together into a self-contained, encapsulated package that can be snapped with other such bundles. So if your requirement is to be able to use widgets, you need an abstraction capable of bundling together MVC, which composes with the rest of the library (i.e. non-widget implementations).

When you have an MVC bundle, what do you need to do with it? Add event listeners and send it commands / requests that have responses, and of course plug it in somewhere in your HTML. *Those are the fundamental requirements to using third-party Javascript components.* They have equal expressive power to React components for the same reason.

In PureScript, components can benefit from strong typing. Moreover, every DOM element can be viewed as a component, which responds to commands / requests, emits events that can be listened to, and is located somewhere in the DOM. Components are in some sense the "assembly language" of the DOM.

If the framework were capable of typing a third-party widget with no expressive power in a way that composes with "the rest of the library", then theoretically you could take advantage of that same power to define, e.g., a reusable look-ahead component, as @natefaubion wanted to do, without any nasty FFI code or third-party library. 

All this is academic right now, but I feel there's a good chance an ace widget will be poorly typed and / or fail to serve the needs of SlamData (but I guess we'll see soon, maybe next week). Beyond that, we have these things called "cells" in SlamData which are intended to be uniform bundles of MVC that can be written externally in JS or other AltJS languages, for which we need some abstraction that will have to somehow interop with Halogen. TBD exactly what that will look like.

**Comment by paf31** [comment]
> Perhaps I should have made the point more forcefully: the free monad approach cannot work at all in Halogen.

Ok, I'll drop it, but I just want to clarify a couple of things.

> So a free monad can only be used for "write-only" DSLs

Absolutely, it doesn't offer anything new or solve a problem, it's just a different way of structuring the input.

> At the very least, res is a superset of req. Also I've never seen real code in which they differ.

The Ace example has two components where `res /= req`. In general, I think we want to be able to decompose a large sum into summands in different ways, which I think is why `Inject1` is so helpful.

> When you have an MVC bundle, what do you need to do with it? Add event listeners and send it commands / requests that have responses.

I really think we need to reconsider (possibly-)effectful signal functions, but I still don't really understand why we need to query the Ace editor. It seems to break our abstraction, and I feel like we can already do the same things with event handlers (instead of "give me your current selection", we subscribe to selection-changed events).

I don't really like the way widgets are modeled in the HTML document right now. I feel like there is possibly a better approach than just sticking a `Widget` in a placeholder. For statically placed widgets, we could model them externally and use the driver function to push their events into the system.

Our two main problems seem to be 

- Being able to treat widgets like pure components
- How to compose signal functions (I would say this is solved, but the solution isn't ideal)

I feel like things are largely solved for the pure case, and the issue is really how to deal with widgets. I don't know what the solution is yet, or even if there is one. 

I still like the option of using the driver function to push effects (including reading from/updating widgets) to the boundary, but that might not work so well with dynamically generated widgets. It's probably still worth considering though, and I might spend some time looking into that.

**Comment by jdegoes** [comment]
> It seems to break our abstraction, and I feel like we can already do the same things with event handlers (instead of "give me your current selection", we subscribe to selection-changed events).

That assumes two things which are not true in the general case:

1. That the component broadcasts all state changes in a way that can be used to reconstruct the internal state of the component. For example, a component which allows a user to draw on the canvas does not broadcast individual painted pixels, nor does Ace broadcast many changes to internal state. A pure Halogen component has the guarantee that its internal state can be duplicated using the initial state and the events coming from the component, but no such guarantee exists for 3rd party components (and, in general, it does not hold for such components).
2. That the client subscribing to the state changes knows how to combine them in such a way to yield an accurate model of the component's internal state. For example, if all you get from the component are key pressed events and keyboard arrow events, then you have to build the state of the document from these events and hope you did so correctly.

Thinking more about (1), one thing I like about the "explicit state" is that it really hammers home the point that the state is purely determined by the initial state and the succession of events produced by the HTML (s, s -> i -> s). It should also be faster since it's one function versus lots of dynamically created functions capturing over state.

But regardless...

> Being able to treat widgets like pure components

I'd say it's more like "being able to integrate all of the impure functionality provided by 3rd party widgets into the semi-pure framework of Halogen". The internal state of foreign widgets cannot be re-created from the events they emit, so if you cannot query them in a pure, type-safe fashion, you have no access to that state. Even in cases where you can re-create it, do you really want to, e.g., recreate the contents of an editor's buffer from primitive operations on it (this is just a random example)?

>  For statically placed widgets, we could model them externally and use the driver function to push their events into the system.

This would be cleaner but I'm not sure it solves the information problem above.

Widgets really want to be something like:

```purescript
type Component p m req res s i = 
  Component { 
    initial :: s, 
    render  :: s -> HTML p i, 
    ask     :: req -> s -> m (Tuple res (s -> s)),
    update  :: i -> m (s -> s) }
```

which composes and can easily be pure, but which doesn't fit so well into Halogen.

One way I can see for making Ace fit Halogen is to effectfully interact with the widget to get the editor's buffer, the caret position, this and that, etc., and stuffing that into a gigantic record (basically, a series of responses to all possible queries). Then on any event that ace exposes, you can include the big, dynamically generated record, which gives you access to the (formerly internal) state of the component.

Another option on some widgets may be trying to expose enough event information that it's actually possible to write a function `s -> i -> s` for some useful amount of state.

**Comment by paf31** [comment]
To a certain extent, that's what the original Ace example did:

- It represented the Ace widget as a pure function from a state (just the buffer contents at that point, a `String`) to a `Widget`. 
- `update` is essentially "diff the state with your internal state, and apply a patch operation", which in this case was just "set the buffer content to this `String`".

This approach is slower obviously, but maybe it's not a bad option if the data is small.

There are conditions on `update` so that you don't run into the "updates being applied twice" issue we saw before. Basically I think the conditions amount to:

- `update` being idempotent: `update newState *> update newState = update newState`
- `update` respects the current state: `getState >>= update` does not change the component's internal state.

In the case where we had to version events to prevent duplicate updates, the second law was being violated: if the current buffer was non-empty and the current state (_last input_) was `ClearText`, then `getState >>= update` would clear the buffer, which affected the internal state of the component.

This is probably only a good option if we decide to go the route of ditching `Component` and developing the `SF` machinery though (@garyb reported good experiences with lenses for example, and there's also `Inject1` now as well).

**Comment by jdegoes** [comment]
> update is essentially "diff the state with your internal state, and apply a patch operation", which in this case was just "set the buffer content to this String".

This would be truly horrifying for a widget allowing the user to paint on a canvas. Only slightly less horrifying for Ace. :smiling_imp: 

> This is probably only a good option if we decide to go the route of ditching Component

Increasingly, I don't see much value in current definition of `Component` vs just wrapping a widget as a `SF` (you can send write-only commands to the component, but you can do that already through SF interface).

**Comment by paf31** [comment]
> Increasingly, I don't see much value in current definition of Component

Agreed, its only value is to hide a few `Either`s. I would gladly get rid of it in 0.5 until we can figure out something better.

**Comment by paf31** [comment]
@jdegoes What do you want to do with this now that #105 and #106 have been merged? I think post-render hooks give us a way to query widgets, and for components, the new approach seems to be plain signal functions and (soon) isomorphisms.
